node/src/components/network.rs

```diff
@@ -437,54 +440,37 @@ where
         count: usize,
         exclude: HashSet<NodeId>,
     ) -> HashSet<NodeId> {
-        let peer_ids = match gossip_target {
-            GossipTarget::Mixed(era_id) => {
-                let (validators, non_validators): (Vec<_>, Vec<_>) = self
-                    .outgoing_manager
-                    .connected_peers()
-                    .filter(|peer_id| !exclude.contains(peer_id))
-                    .partition(|node_id| {
-                        self.outgoing_limiter.is_validator_in_era(era_id, node_id)
-                    });
-
-                non_validators
-                    .choose_multiple(rng, count)
-                    .interleave(validators.iter().choose_multiple(rng, count))
-                    .take(count)
-                    .copied()
-                    .collect()
-            }
-            GossipTarget::NonValidators(era_id) => {
-                self.outgoing_manager
-                    .connected_peers()
-                    .filter(|peer_id| {
-                        if exclude.contains(peer_id) {
-                            return false;
-                        }
-                        // If the peer isn't a validator, include it.
-                        !self.outgoing_limiter.is_validator_in_era(era_id, peer_id)
-                    })
-                    .choose_multiple(rng, count)
-            }
-            GossipTarget::All => self
-                .outgoing_manager
-                .connected_peers()
-                .filter(|peer_id| !exclude.contains(peer_id))
-                .choose_multiple(rng, count),
-        };
+        let is_validator_in_era =
+            |era: EraId, peer_id: &NodeId| self.outgoing_limiter.is_validator_in_era(era, peer_id);
+        let peer_ids = choose_gossip_peers(
+            rng,
+            gossip_target,
+            count,
+            exclude.clone(),
+            self.outgoing_manager.connected_peers(),
+            is_validator_in_era,
+        );
 
         // todo!() - consider sampling more validators (for example: 10%, but not fewer than 5)
 
         if peer_ids.len() != count {
-            // TODO - set this to `warn!` once we are normally testing with networks large enough to
-            //        make it a meaningful and infrequent log message.
-            trace!(
-                our_id=%self.context.our_id(),
-                wanted = count,
-                selected = peer_ids.len(),
-                "could not select enough random nodes for gossiping, not enough non-excluded \
-                outgoing connections"
-            );
+            let connected = self.outgoing_manager.connected_peers().count();
+            let not_excluded = self
+                .outgoing_manager
+                .connected_peers()
+                .filter(|peer_id| !exclude.contains(peer_id))
+                .count();
+            if not_excluded > 0 {
```

Comment:
So if everyone is excluded we don't want to see the debug message?

Reply:
Yes, since this happens very frequently when using nctl defaults.